lib/elixir/lib/calendar/date/range.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Date.Range do
+  @moduledoc """
+  Defines a range of dates.
+
+  A range of dates represents a discrete number of dates where
+  the first and last values are dates with matching calendars.
+
+  Ranges of dates can be either increasing (`first <= last`) or
+  decreasing (`first > last`). They are also always inclusive.
+
+  A range of dates implements the `Enumerable` protocol, which means
+  functions in the `Enum` module can be used to work with
+  ranges:
+
+      iex> range = Date.range(~D[2001-01-01], ~D[2002-01-01])
+      iex> Enum.count(range)
+      366
+      iex> Enum.member?(range, ~D[2001-02-01])
+      true
+      iex> Enum.reduce(range, 0, fn(_date, acc) -> acc - 1 end)
+      -366
+
+  """
+
+  defstruct [:first, :last, :kind, :first_rata_die, :last_rata_die]
+
+  def new(%{calendar: fc} = first, %{calendar: lc} = last) when fc == lc do
+    %Date.Range{
+      first: first,
+      last: last,
+      first_rata_die: Date.to_rata_die(first),
+      last_rata_die: Date.to_rata_die(last),
+      kind: get_range_kind(first, last),
```

Comment:
If we remove it, we need to do the comparison in many places. Do we want that, or do we prefer to keep it in one place?